witchcraft/internal/tcpjson/tcp_logger.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpjson
+
+import (
+	"encoding/json"
+	"io"
+	"net"
+	"sync"
+
+	werror "github.com/palantir/witchcraft-go-error"
+)
+
+var _ io.Writer = (*TCPWriter)(nil)
+
+const ErrWriterClosed = "writer is closed"
+
+// TCPWriter writes logs to a TCP socket and wraps them with envelope metadata.
+type TCPWriter struct {
+	metadata LogEnvelopeMetadata
+	provider ConnProvider
+
+	// closedChan is used to signal that the writer is shutting down
+	closedChan chan struct{}
+
+	mu   sync.RWMutex // guards conn below
+	conn net.Conn
+}
+
+func NewTCPWriter(metadata LogEnvelopeMetadata, provider ConnProvider) *TCPWriter {
+	return &TCPWriter{
+		metadata:   metadata,
+		provider:   provider,
+		closedChan: make(chan struct{}),
+		conn:       nil,
+	}
+}
+
+func (d *TCPWriter) Write(logPayload []byte) (int, error) {
+	if d.closed() {
+		return 0, werror.Error(ErrWriterClosed)
+	}
+	envelopeToWrite := SlsEnvelopeV1{
+		Type:           "envelope.1",
+		Deployment:     d.metadata.Deployment,
+		Environment:    d.metadata.Environment,
+		EnvironmentID:  d.metadata.EnvironmentID,
+		Host:           d.metadata.Host,
+		NodeID:         d.metadata.NodeID,
+		Service:        d.metadata.Service,
+		ServiceID:      d.metadata.ServiceID,
+		Stack:          d.metadata.Stack,
+		StackID:        d.metadata.StackID,
+		Product:        d.metadata.Product,
+		ProductVersion: d.metadata.ProductVersion,
+		Payload:        logPayload,
+	}
+	b, err := json.Marshal(envelopeToWrite)
```

Comment:
Should we use `safejson.Marshal` here? I seem to recall that the default adds a newline and doesn't handle certain encodings etc. (which was our motivation for having `safejson`)

Reply:
unfortunately not because `safejson.Marshal` uses `json.Encoder` under the hood and then strips off the `\n` char that we need to keep. I have since replaced this entirely as well.